LIPs/LIP-service-registry.md

```diff
@@ -0,0 +1,88 @@
+
+In proposed [future architectures](https://github.com/livepeer/go-livepeer/issues/430) that do not rely on a libp2p based network, but rather an HTTP request-reply workflow using known service endpoints, transcoders and broadcasters will require an alternative method to establish off-chain connections with each other. Additionally, ideally broadcasters would not have to publish their own connection endpoints to the world as this would increase the operational and security overhead for a broadcaster node operator to defend against  DDOS attacks. Instead, transcoders are likely better candidates for publishing connection endpoints to the world as there is a higher expectation for operational and security capatabilities for transcoder node operators.
+
+Possible approaches for transcoders to advertise connection endpoints include DHT based discovery, a global registration HTTP endpoint and a smart contract registry. The downsides of DHT based discovery center around current unknowns around stability and engineering overhead. The downsides of a global registration HTTP endpoint center around the reliance on a centralized service. Consequently, a smart contract registry of service endpoints that transcoders can be use to publish an endpoint that broadcasters can send requests to appears to be a reasonable approach at this time.
+
+## Specification
+
+Create the contract ServiceRegistry with the following interface. A full implementation is provided in a later section of this proposal:
+
+```
+interface IServiceRegistry {
+    event ServiceURIUpdate(address indexed addr, string serviceURI);
+
+    function setServiceURI(string _serviceURI) external;
+    function getServiceURI(address _addr) public view returns (string);
+}
+```
+
+The contract uses a `Record` struct to store the service URI thus allowing for the addition of other state variables in the future if necessary.
+
+The contract uses a state variable `mapping (address => Record) private records` to manage the registry records.
+
+The function `setServiceURI(string _serviceURI)` does the following:
+- Store `_serviceURI` as the service URI endpoint for the caller that can be used to send requests to the caller off-chain
+- This function will only set the service URI endpoint for the caller and no other accounts
+- Successful execution of this function will emit the `ServiceURIUpdate()` event that will include the address of the caller and its new 
+service URI. The address is indexed thus allowing clients to filter for `ServiceURIUpdate()` events specifically for changes from a specific caller address
+
+The function `getServiceURI(address _addr)` does the following:
+- Return the service URI endpoint stored for the caller
```

Comment:
> Return the service URI endpoint stored for the caller

Should this say "service URI endpoint stored for the address" ?

Random thought about building in some semblance of protection. Since `getServiceURI` will be a function that's free to read, could we extend it to limit reads to only callers that have an on-chain job assigned to the target address? So if a node tries to call `getServiceURI(jobID, address)` without being responsible for the job, or being assigned the given address, it wouldn't get the URI back.

This wouldn't stop someone who *really* wanted to extract addresses, but it'd discourage drive-by poking. Might straitjacket us later on though, if we ever need to do arbitrary lookups.

Reply:
Yeah good catch - `getServiceURI(address _addr)` should return the service URI endpoint stored for the address `_addr` (and not for the caller).

Although we could place restrictions on the service URI getter to revert/fail if the caller shouldn't be allowed to know the service URI, I'm not sure if introducing the additional logic which would make the JobsManager contract a dependency for this contract (one nice property of this proposed contract right now is that it has 0 other contract dependencies) is worth it given that it is pretty easy for anyone to extract the service URI from the contract's storage using the `eth_storageAt` RPC method (https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getstorageat) - anyone could precalculate the position in storage that the service URI would live at and then use the RPC method to extract it.